R/Backend-class.R

```diff
@@ -266,7 +279,57 @@ setGeneric(
 #' @noRd
 setGeneric(
     "backendWriteSpectra",
-    def=function(object, file, spectra, spectraData, ..., BPPARAM=bpparam())
+    def=function(object, spectra, spectraData, ..., BPPARAM=bpparam())
         standardGeneric("backendWriteSpectra"),
     valueClass="Backend"
 )
+
+#' Apply a function to each spectrum (after applying eventual lazy evaluation
+#' steps) and return its result. If `FUN = NULL` the function should return
+#' the `list` of `Spectrum` objects.
+#'
+#' @inheritParams backendInitialize
+#'
+#' @param FUN either a function or the name of a function to be applied to
+#'     each spectrum.
+#'
+#' @param ... additional arguments to `FUN`
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+setGeneric(
+    "backendSpectrapply",
+    def = function(object, spectraData, FUN = NULL, ..., BPPARAM = bpparam())
+        standardGeneric("backendSpectrapply"),
+    valueClass = "list"
+)
+
+#' Add a processing step to the backend. By default it will be added to the
+#' processingQueue of the backend, but the method could also be overwritten to
+#' directly apply the processing step.
+#'
+#' @inheritParams backendInitialize
+#'
+#' @param procStep `ProcessingStep`
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+setGeneric(
+    "backendAddProcessing",
+    def = function(object, spectraData, procStep, ...)
+        standardGeneric("backendAddProcessing"),
+    valueClass = "Backend"
+)
+
+setMethod(
+    "backendAddProcessing", "Backend", function(object, spectraData,
+                                                procStep, ...) {
+        if (missing(procStep) || !inherits(procStep, "ProcessingStep"))
+            stop("'procStep' parameter is required and is expected to be of",
+                 " type 'ProcessingStep'")
```

Comment:
The signature could be `c(object="Backend", spectraData="DataFrame", procStep="ProcessingStep")` and we could remove the `if(missing(procStep) ...` (but we lose a clear error message, but because the user will never call any `backend...` methods that doesn't matter).